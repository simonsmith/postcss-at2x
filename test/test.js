import postcss from 'postcss';
import fs from 'fs';
import path from 'path';
import plugin from '../lib/';

function readFixture(filename) {
  return fs.readFileSync(path.join('test/fixtures', filename), 'utf-8');
}

function runTest(input, opts, done) {
  input = readFixture(`${input}`);

  postcss([plugin(opts)]).process(input).then((result) => {
    expect(result.css).toMatchSnapshot();
    expect(result.warnings()).toHaveLength(0);
    done();
  }).catch(done);
}

describe('postcss-at2x', () => {
  it('should add device-pixel-ratio rules', (done) => {
    runTest('at2x.css', {}, done);
  });

  it('should allow retina image scoped to media query', (done) => {
    runTest('scoped.css', {}, done);
  });

  it('should allow a custom identifier for retina file names', (done) => {
    runTest('identifier.css', {identifier: '-retina'}, done);
  });

  it('should allow mulitple background images to be replaced', (done) => {
    runTest('multibg.css', {}, done);
  });

  it('should process image and add background size', (done) => {
    runTest('read-background-size.css', {detectImageSize: true}, done);
  });

  it('should resolve image path with custom function', (done) => {
    runTest('resolve-image-path.css', {
      detectImageSize: true,
      resolveImagePath: (value, source) => {
        expect(typeof value).toBe('string');
        expect(source).toHaveProperty('input.css');
        return path.resolve(process.cwd(), value);
      }}, done);
  });

  it('should have proper arguments for resolve image path function', (done) => {
    const spy = jest.fn();
    runTest('resolve-image-path.css', {detectImageSize: true, resolveImagePath: spy}, done);
    const spyCall = spy.mock.calls[0];
    expect(spyCall[0]).toEqual('./test/fixtures/images/cat.jpg');
    expect(spyCall[1]).toMatchSnapshot();
  });
});
