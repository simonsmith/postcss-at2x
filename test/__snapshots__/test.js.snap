// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcss-at2x should add device-pixel-ratio rules 1`] = `
"/**
 * Multiple background values being set at once
 */
.logo {
  background: red url('/public/images/logo.png') no-repeat 0 0;
  background-size: auto auto;
}

/**
 * Just an image
 */
.user {
  background-image: url('/public/images/site.png');
}

/**
 * Keyword doesn't have to go at the end
 */
.user {
  background: transparent url('/public/images/site.png') repeat 0 0;
}

/**
 * Ignoring declarations without at-2x
 */
.avatar {
  background-image: url('/public/images/site.png');
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .logo {
    background: red url('/public/images/logo@2x.png') no-repeat 0 0;
    background-size: auto auto;
  }
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .user {
    background-image: url('/public/images/site@2x.png');
  }
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .user {
    background: transparent url('/public/images/site@2x.png') repeat 0 0;
  }
}
"
`;

exports[`postcss-at2x should allow a custom identifier for retina file names 1`] = `
".user {
  background-image: url('/public/images/site.png');
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .user {
    background-image: url('/public/images/site-retina.png');
  }
}
"
`;

exports[`postcss-at2x should allow mulitple background images to be replaced 1`] = `
".multi {
  background: url(http://example.com/image.png),
              url(/public/images/cool.png),
              url(http://example.com/flowers-pattern.jpg);
}

.multi {
  background-image: url(http://example.com/image.png),
                    linear-gradient(to right, rgba(255, 255, 255, 0),  rgba(255, 255, 255, 1)),
                    url(/public/images/cool.png);
}

@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {

  .multi {
    background: url(http://example.com/image.png),
              url(/public/images/cool@2x.png),
              url(http://example.com/flowers-pattern@2x.jpg);
  }
}

@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {

  .multi {
    background-image: url(http://example.com/image.png),
                    linear-gradient(to right, rgba(255, 255, 255, 0),  rgba(255, 255, 255, 1)),
                    url(/public/images/cool@2x.png);
  }
}
"
`;

exports[`postcss-at2x should allow retina image scoped to media query 1`] = `
".Logo {
  background-image: url(\\"../images/logo-1.png\\");
}

@media (min-width: 600px) {
  .Logo {
    background-image: url(\\"../images/logo-2.png\\");
  }
}

.Logo {
  background-image: url(\\"../images/logo-3.png\\");
}

@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {

  .Logo {
    background-image: url(\\"../images/logo-1@2x.png\\");
  }
}

@media (min-width: 600px) and (min-device-pixel-ratio: 1.5), (min-width: 600px) and (min-resolution: 144dpi), (min-width: 600px) and (min-resolution: 1.5dppx) {

  .Logo {
    background-image: url(\\"../images/logo-2@2x.png\\");
  }
}

@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {

  .Logo {
    background-image: url(\\"../images/logo-3@2x.png\\");
  }
}
"
`;

exports[`postcss-at2x should have proper arguments for resolve image path function 1`] = `
Object {
  "end": Object {
    "column": 64,
    "line": 2,
  },
  "input": Input {
    "css": ".cat {
  background-image: url('./test/fixtures/images/cat.jpg') at-2x;
}
",
    "id": "<input css 7>",
  },
  "start": Object {
    "column": 3,
    "line": 2,
  },
}
`;

exports[`postcss-at2x should process image and add background size 1`] = `
".cat {
  background-image: url('./test/fixtures/images/cat.jpg');
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .cat {
    background-image: url('./test/fixtures/images/cat@2x.jpg');
    background-size: 200px 286px;
  }
}
"
`;

exports[`postcss-at2x should resolve image path with custom function 1`] = `
".cat {
  background-image: url('./test/fixtures/images/cat.jpg');
}
@media (min-device-pixel-ratio: 1.5), (min-resolution: 144dpi), (min-resolution: 1.5dppx) {
  .cat {
    background-image: url('./test/fixtures/images/cat@2x.jpg');
    background-size: 200px 286px;
  }
}
"
`;
